tests/components/tasmota/test_discovery.py

```diff
@@ -0,0 +1,372 @@
+
+    # Trigger an exception when the entity is added
+    with patch(
+        "hatasmota.discovery.get_device_config_helper",
+        return_value=object(),
+    ):
+        async_fire_mqtt_message(hass, f"{DEFAULT_PREFIX}/{mac}/config", data)
+        await hass.async_block_till_done()
+
+    # Verify device entry is not created
+    device_entry = device_reg.async_get_device(set(), {("mac", mac)})
+    assert device_entry is None
+    assert (
+        "Exception in async_discover_device when dispatching 'tasmota_discovery_device'"
+        in caplog.text
+    )
+
+    async_fire_mqtt_message(hass, f"{DEFAULT_PREFIX}/{mac}/config", data)
+    await hass.async_block_till_done()
+
+    # Verify device entry is created
+    device_entry = device_reg.async_get_device(set(), {("mac", mac)})
+    assert device_entry is not None
+
+
+async def test_device_remove(
+    hass, mqtt_mock, caplog, device_reg, entity_reg, setup_tasmota
+):
+    """Test removing a discovered device."""
+    config = copy.deepcopy(DEFAULT_CONFIG)
+    mac = config["mac"]
+
+    async_fire_mqtt_message(
+        hass,
+        f"{DEFAULT_PREFIX}/{mac}/config",
+        json.dumps(config),
+    )
+    await hass.async_block_till_done()
+
+    # Verify device entry is created
+    device_entry = device_reg.async_get_device(set(), {("mac", mac)})
+    assert device_entry is not None
+
+    async_fire_mqtt_message(
+        hass,
+        f"{DEFAULT_PREFIX}/{mac}/config",
+        "",
+    )
+    await hass.async_block_till_done()
+
+    # Verify device entry is removed
+    device_entry = device_reg.async_get_device(set(), {("mac", mac)})
+    assert device_entry is None
+
+
+async def test_device_remove_stale(hass, mqtt_mock, caplog, device_reg, setup_tasmota):
+    """Test removing a stale (undiscovered) device does not throw."""
+    mac = "00000049A3BC"
+
+    config_entry = hass.config_entries.async_entries("tasmota")[0]
+
+    # Create a device
+    device_reg.async_get_or_create(
+        config_entry_id=config_entry.entry_id,
+        connections={("mac", mac)},
+    )
+
+    # Verify device entry was created
+    device_entry = device_reg.async_get_device(set(), {("mac", mac)})
+    assert device_entry is not None
+
+    # Remove the device
+    device_reg.async_remove_device(device_entry.id)
+
+    # Verify device entry is removed
+    device_entry = device_reg.async_get_device(set(), {("mac", mac)})
+    assert device_entry is None
+
+
+async def test_device_rediscover(
+    hass, mqtt_mock, caplog, device_reg, entity_reg, setup_tasmota
+):
+    """Test removing a device."""
+    config = copy.deepcopy(DEFAULT_CONFIG)
+    mac = config["mac"]
+
+    async_fire_mqtt_message(
+        hass,
+        f"{DEFAULT_PREFIX}/{mac}/config",
+        json.dumps(config),
+    )
+    await hass.async_block_till_done()
+
+    # Verify device entry is created
+    device_entry1 = device_reg.async_get_device(set(), {("mac", mac)})
+    assert device_entry1 is not None
```

Comment:
Codacy found an issue: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=6285789)